Fix create home dir fallback

kustodian opened the pull request:
##### SUMMARY
When a user home dir is created by the `user` module (instead by `useradd` command which is run by this module), the home dir will now be created with `0700` mode (it was created with `0755` before which is far to open for a home dir). Also, now it will create home even if `/etc/skel` dir doesn't exist (which was a bug).

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
user.py

##### ADDITIONAL INFORMATION
This PR also fixes one edge case bug in which after a user is deleted, its home dir is created. To give an example of what I'm talking about is if the same user exists as a local user and as a central system (remote) user (e.g. AD/FreeIPA/LDAP, etc). If we want to delete the local user using this module, after the user is deleted, a home dir will be created for the remote user, which is a bug and it shouldn't happen. See my comment in the code.

I also added tests for creating the user home using this fallback mechanism, especially since it's a rare code path.

kustodian committed:
Fix create home dir fallback

When a user home dir is not created with `useradd`, the home dir will now
be created with umask from /etc/login.defs. Also fixed a bug in which
after a local user is deleted, and the same user exists in the central
user management system, the module would create that user's home.

kustodian commented on lib/ansible/modules/system/user.py:
```diff
@@ -2638,7 +2642,7 @@ def main():
         result['stderr'] = err
 
-    if user.user_exists():
+    if user.user_exists() and user.state == 'present':
```
We should only fetch user information with state `present`, there is no need to fetch user info if we are deleting a user.

samdoran replied:
If `user.state == 'absent'` the user account will be removed, which means `user.user_exists()` will be `False`. No need for the additional condition.

kustodian replied:
But it doesn't hurt to have this check, because we had an issue because of it. We had local users on our hosts and then we added those hosts into a FreeIPA domain. The FreeIPA domain had the same exact users that were created locally before, so when we deleted our local users using this module, what happened was that this condition returned `true` that the user did exist (because it found a user on the FreeIPA domain) and then it created that user's home dir.

For example if there is a `kustodian` local user and the `kustodian` user exists on the FreeIPA domain (or any central user management system), when you run:

```
- user:
    name: kustodian
    state: absent
```
This would delete the local `kustodian` user and then it would create the remote `kustodian` user home dir, which in our case was `/home/ipa.lan/kustodian`. Which is not what you expect from the `user` module to do when it deletes a user. Also it created the home dirs with `755`, which we also don't want (`useradd` creates them with 700). I know this is an edge case, but its something that can happen (it happened to us) and this additional condition solves the problem.

samdoran replied:
I see. Thank you for that great explanation. Since it doesn’t hurt to have it plus fixes the scenario you describe, we should keep it.

kustodian commented on lib/ansible/modules/system/user.py:
```diff
@@ -1001,9 +1001,13 @@ def create_homedir(self, path):
+                try:
+                    os.makedirs(path)
+                except OSError as e:
+                    self.module.exit_json(failed=True, msg="%s" % to_native(e))
```
This was an indentation bug, because it doesn't make sense to create the home dir, if it already exists, and it also doesn't make sense not to create it if `/etc/skell` doesn't exists, which was the case before.

samdoran replied:
Good catch.

samdoran commented on lib/ansible/modules/system/user.py:
```diff
@@ -1001,9 +1001,13 @@ def create_homedir(self, path):
+                    self.module.exit_json(failed=True, msg="%s" % to_native(e))
             try:
-                os.makedirs(path)
+                os.chmod(path, 0o700)
```
I don't believe we should be hard coding permissions here. We should use the system default value from `/etc/login.defs` or add this as a module parameter the defaults to the system value.

kustodian replied:
Good call. I'll check the value in `/etc/login.defs` and possible add a new parameter.

kustodian replied:
In the end I decided not to add a new parameter, because it would just complicate things in situations when a user home is being created with `useradd`. Now it tries to read `UMASK` from `/etc/login.defs` and if it it doesn't find it, it just uses the default.

samdoran reviewed (changes requested):
I like the current solution. Regarding the tests, should they check that the mode on the created home dir matches what it is `/etc/login.defs`?

kustodian commented:
I thought about that, but I didn't think of an easy way of testing that in
Ansible, so I just said to myself that it's probably not that important,
especially since this is a rare use case. I can give it a another try if
you think it's important, or if you have a tip how to best do it, I'm all
ears. For now my idea is to call python directly to convert what I got from
UMASK into mode.

kustodian commented:
@samdoran I added a test for the home directory mode and after I added it I noticed I had a bug where mode wasn't set if the home dir was created by copying `/etc/ske`, so I fixed it and now all should be good to go (if all the tests pass).

samdoran commented on lib/ansible/modules/system/user.py:
```diff
@@ -1001,11 +1002,22 @@ def create_homedir(self, path):
+                    self.module.exit_json(failed=True, msg="%s" % to_native(e))
+            # get umask from /etc/login.defs and set correct home mode
+            if os.path.exists(self.LOGIN_DEFS):
+                for line in open(self.LOGIN_DEFS).readlines():
```
Use a context manager with this read operation to ensure the file handler is closed as soon as it is no longer needed rather than relying on successful program termination.
```suggestion
                with open(self.LOGIN_DEFS, 'r') as f:
                    for line in f.readlines():
```

kustodian replied:
Thanks, I have no idea why I didn't use it, because I remember doing it :D

kustodian replied:
While I was fixing this, I updated all file operations to use `with` because it was all over the place.

kustodian commented:
@samdoran would you be able to get this one merged, since all the requested changes are now resolved :)

samdoran commented:
Please create a changelog fragment. See [fragments](https://github.com/ansible/ansible/tree/stable-2.7/changelogs/fragments) for examples.

kustodian commented:
I added the fragment, squashed all commits into one and rebased everything, so once the tests pass (and they should because I didn't change anything), this should be good to go. Thanks for your help @samdoran  :)

kustodian commented:
I had to force push one more time, because the CI was stale. Now all should be good and ready for merging.

samdoran commented:
rebuild_merge
